### PR TITLE
MAINT: close assorted small issues

### LINKED ISSUE #4110: KeyError?
https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tools/decorators.py#L103

```
            try:
                _cache[name] = _cachedval
            except KeyError:
                setattr(_cache, name, _cachedval)
```

What's the scenario in which a KeyError is raised here?

### LINKED ISSUE #4349: commented-out code in tools.tests
epsilon is defined and then not used, with a comment that it is not precise enough.  So... what purpose does it serve?
https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tools/tests/test_numdiff.py#L197

Are these four tests ever going to be un-commented-out?  Could they at least be xfailed?https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tools/tests/test_tools.py#L312

### LINKED ISSUE #4419: #delete when testing is finished
sandbox.stats.diagnostic line 1414 just above `class StatTestMC`

### LINKED ISSUE #4451: unclosed paren in adfvalues
https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tsa/adfvalues.py#L103

```
# the left tail approximation is
# p = norm.cdf(d_0 + d_1*log(abs(z)) + d_2*log(abs(z))**2 + d_3*log(abs(z))**3
# there is no Z-min, ie., it is well-behaved in the left tail
```

please advise.

### LINKED ISSUE #5276: Is overwriting approx_hess_cs.__doc__ intentional?

### LINKED ISSUE #5278: Unclear expression in iolib.foreign
```
            self._str = value-offset == 1 and \
                    '.' or ('.' + chr(value-offset+96))
```

### LINKED ISSUE #5687: BUG: undefined name in grouputils.dummy_sparse
The content of the function is:

```
def dummy_sparse(groups):
    from scipy import sparse

    indptr = np.arange(len(groups)+1)
    data = np.ones(len(groups), dtype=np.int8)
    indi = sparse.csr_matrix((data, g, indptr))

    return indi
```

But the `g` variable is undefined.  I'm _pretty_ sure that should be `groups` instead of `g`.  @josef-pkt can you confirm this guess?  If so, would the example in the docstring make for a valid test case if I push a PR fixing this?

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #4451, closes #4419, closes #4349, closes #4110, closes #5278, closes #5276, closes #5687
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message). 

This is a hodge-podge.  I'll be happy to break it up into any requested subsets, but didn't want to make a bunch of tiny PRs right now.

### CODE REVIEW

File: statsmodels/discrete/tests/test_discrete.py [hunk 1]
```diff
@@ -1116,11 +1116,17 @@ def test_conf_int(self):
...
                             DECIMAL_2)
 
+    @pytest.mark.xfail(reason="No idea!  But until this test is actually "
+                              "written, it is considered broken.",
+                       strict=True, raises=NotImplementedError)
     def test_predict(self):
-        pass
+        raise NotImplementedError
 
+    @pytest.mark.xfail(reason="No idea!  But until this test is actually "
```
Comment (bashtage) [inline statsmodels/discrete/tests/test_discrete.py]: Could you use a simpler reason, e.g., "test has not been implemented for this class"?  A one-line reason seems to be enough.  Also, would be great to use the same simple reason everywhere so these are coalesced in pytest reports.

Comment (jbrockmendel) [reply statsmodels/discrete/tests/test_discrete.py]: Sure.  FWIW the "No idea!" is in place for several similar xfails, so you can search on that if you're looking for where else this has been done.

File: statsmodels/tools/tests/test_numdiff.py [hunk 2]
```diff
@@ -194,7 +194,6 @@ def test_grad_fun1_fdc(self):
             gtrue = self.gradtrue(test_params)
             fun = self.fun()
 
-            epsilon = 1e-6  #default epsilon 1e-6 is not precise enough
```
Comment (bashtage) [inline statsmodels/tools/tests/test_numdiff.py]: Leave in comment

File: statsmodels/tools/tests/test_tools.py [hunk 3]
```diff
@@ -290,60 +290,88 @@ def test_structarray1d_drop(self):
...
 
-# comment out until we have type coercion
-#    def test_array2d(self):
-#        des = np.column_stack((self.des, self.instr, self.des))
-#        des = tools.categorical(des, col=2)
-#        assert_array_equal(des[:,-5:], self.dummy)
-#        assert_equal(des.shape[1],10)
+    @pytest.mark.xfail(reason="No idea!  But xfailing instead of leaving "
+                              "this commented out with the comment "
+                              "'comment out until we have type coercion'")
```
Comment (bashtage) [inline statsmodels/tools/tests/test_tools.py]: `'comment out` -> `commented out`

File: statsmodels/tools/tests/test_tools.py [hunk 4]
```diff
@@ -406,17 +434,30 @@ def test_structarray1d_drop(self):
...
         assert_equal(len(dum.dtype.names), 5)
 
+    @pytest.mark.xfail(reason="No idea!  But until this test is actually "
+                              "written, it is considered broken.",
+                       strict=True, raises=NotImplementedError)
     def test_arraylike2d(self):
-        pass
+        raise NotImplementedError
 
+    @pytest.mark.xfail(reason="No idea!  But until this test is actually "
```
Comment (bashtage) [inline statsmodels/tools/tests/test_tools.py]: Friendly message for NotImplementedError please, same everywhere.

**Comment by jbrockmendel** [comment]
Addressed+green.  I think I squashed down a couple commits, still have 8 since they are each issue-specific

**Pull request merged**